src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Documents/VisualStudioEditorDocumentManager.cs

```diff
@@ -26,55 +30,59 @@ internal class VisualStudioEditorDocumentManager : EditorDocumentManagerBase
 
         public VisualStudioEditorDocumentManager(
             ForegroundDispatcher foregroundDispatcher,
+            JoinableTaskContext joinableTaskContext,
             FileChangeTrackerFactory fileChangeTrackerFactory,
             IVsRunningDocumentTable runningDocumentTable,
             IVsEditorAdaptersFactoryService editorAdaptersFactory)
-            : base(foregroundDispatcher, fileChangeTrackerFactory)
+            : base(foregroundDispatcher, joinableTaskContext, fileChangeTrackerFactory)
         {
-            if (runningDocumentTable == null)
+            if (runningDocumentTable is null)
             {
                 throw new ArgumentNullException(nameof(runningDocumentTable));
             }
 
-            if (editorAdaptersFactory == null)
+            if (editorAdaptersFactory is null)
             {
                 throw new ArgumentNullException(nameof(editorAdaptersFactory));
             }
 
-            if (foregroundDispatcher == null)
-            {
-                throw new ArgumentNullException(nameof(foregroundDispatcher));
-            }
-
-            if (fileChangeTrackerFactory == null)
+            if (fileChangeTrackerFactory is null)
             {
                 throw new ArgumentNullException(nameof(fileChangeTrackerFactory));
             }
 
             _runningDocumentTable = (IVsRunningDocumentTable4)runningDocumentTable;
             _editorAdaptersFactory = editorAdaptersFactory;
             
-            var hr = runningDocumentTable.AdviseRunningDocTableEvents(new RunningDocumentTableEventSink(this), out _);
+            var hr = runningDocumentTable.AdviseRunningDocTableEvents(new RunningDocumentTableEventSink(this, foregroundDispatcher), out _);
             Marshal.ThrowExceptionForHR(hr);
 
             _documentsByCookie = new Dictionary<uint, List<DocumentKey>>();
             _cookiesByDocument = new Dictionary<DocumentKey, uint>();
         }
 
-        protected override ITextBuffer GetTextBufferForOpenDocument(string filePath)
+        protected override async Task<ITextBuffer> GetTextBufferForOpenDocumentAsync(string filePath)
         {
             if (filePath == null)
             {
                 throw new ArgumentNullException(nameof(filePath));
             }
 
+            ForegroundDispatcher.AssertForegroundThread();
```

Comment:
Hmm, any reason for this path to require that it's on the foreground thread to start? I imagine this entire call path may want to be on the UI thread since it's interacting with the running doc table.